iroh/src/node.rs

```diff
@@ -963,6 +965,147 @@ impl<D: BaoStore> RpcHandler<D> {
         rx.into_stream().map(BlobAddPathResponse)
     }
 
+    fn doc_import_file(
+        self,
+        msg: DocImportFileRequest,
+    ) -> impl Stream<Item = DocImportFileResponse> {
+        // provide a little buffer so that we don't slow down the sender
+        let (tx, rx) = flume::bounded(32);
+        let tx2 = tx.clone();
+        self.rt().local_pool().spawn_pinned(|| async move {
+            if let Err(e) = self.doc_import_file0(msg, tx).await {
+                tx2.send_async(DocImportProgress::Abort(e.into()))
+                    .await
+                    .ok();
+            }
+        });
+        rx.into_stream().map(DocImportFileResponse)
+    }
+
+    async fn doc_import_file0(
+        self,
+        msg: DocImportFileRequest,
+        progress: flume::Sender<DocImportProgress>,
+    ) -> anyhow::Result<()> {
+        use iroh_bytes::store::ImportMode;
+        use std::collections::BTreeMap;
+
+        let progress = FlumeProgressSender::new(progress);
+        let names = Arc::new(Mutex::new(BTreeMap::new()));
+        // convert import progress to provide progress
+        let import_progress = progress.clone().with_filter_map(move |x| match x {
+            ImportProgress::Found { id, name } => {
+                names.lock().unwrap().insert(id, name);
+                None
+            }
+            ImportProgress::Size { id, size } => {
+                let name = names.lock().unwrap().remove(&id)?;
+                Some(DocImportProgress::Found { id, name, size })
+            }
+            ImportProgress::OutboardProgress { id, offset } => {
+                Some(DocImportProgress::Progress { id, offset })
+            }
+            ImportProgress::OutboardDone { hash, id } => {
+                Some(DocImportProgress::IngestDone { hash, id })
+            }
+            _ => None,
+        });
+        let DocImportFileRequest {
+            doc_id,
+            author_id,
+            key,
+            path: root,
+            in_place,
+        } = msg;
+        // Check that the path is absolute and exists.
+        anyhow::ensure!(root.is_absolute(), "path must be absolute");
+        anyhow::ensure!(
+            root.exists(),
+            "trying to add missing path: {}",
+            root.display()
+        );
+
+        let import_mode = match in_place {
+            true => ImportMode::TryReference,
+            false => ImportMode::Copy,
+        };
+
+        let (temp_tag, size) = self
+            .inner
+            .db
+            .import_file(root, import_mode, BlobFormat::Raw, import_progress)
+            .await?;
+
+        let hash_and_format = temp_tag.inner();
+        let HashAndFormat { hash, .. } = *hash_and_format;
```

Comment:
@Frando @rklaehn Is this enough to handle tags correctly? Do I need to do more than just let the `temp_tag` drop?

Reply:
I think this is the case, yes, see the drop impl for  `TempTag` which informs the "liveness tracker" of the store that the tag is dropped.

Reply:
Yes, I think this is correct. You are keeping the temp tag around until you store it in the document. As soon as it is in the document, it will be protected by the document.

If you want to make it more explicit, you can explicitly call drop on the temp tag after the add to the doc.